src/domain/datasets/services/src/use_cases/update_version_file_use_case_impl.rs

```diff
@@ -0,0 +1,227 @@
+use kamu_core::{
+    DatasetRegistry,
+    FileUploadLimitConfig,
+    GetDataOptions,
+    PushIngestDataError,
+    PushIngestDataUseCase,
+    PushIngestError,
+    QueryService,
+};
+use kamu_datasets::{
+    ContentArgs,
+    ExtraDataFields,
+    FileVersion,
+    UpdateVersionFileResult,
+    UpdateVersionFileUseCase,
+    UpdateVersionFileUseCaseError,
+    UploadTooLargeError,
+    VersionedFileEntity,
+};
+use serde_json::Value;
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[component]
+#[interface(dyn UpdateVersionFileUseCase)]
+pub struct UpdateVersionFileUseCaseImpl {
+    dataset_registry: Arc<dyn DatasetRegistry>,
+    upload_config: Arc<FileUploadLimitConfig>,
+    push_ingest_data_use_case: Arc<dyn PushIngestDataUseCase>,
+    query_svc: Arc<dyn QueryService>,
+}
+
+impl UpdateVersionFileUseCaseImpl {
+    async fn get_latest_version(
+        &self,
+        dataset_ref: &odf::DatasetRef,
+    ) -> Result<(FileVersion, odf::Multihash), InternalError> {
+        // TODO: Consider retractions / corrections
+        let query_res = self
+            .query_svc
+            .tail(dataset_ref, 0, 1, GetDataOptions::default())
+            .await
+            .int_err()?;
+
+        let Some(df) = query_res.df else {
+            return Ok((0, query_res.block_hash));
+        };
+
+        let last_version = df
+            .select_columns(&["version"])
+            .int_err()?
+            .collect_scalar::<datafusion::arrow::datatypes::Int32Type>()
+            .await
+            .int_err()?;
+
+        let last_version = FileVersion::try_from(last_version.unwrap_or(0)).unwrap();
+
+        Ok((last_version, query_res.block_hash))
+    }
+
+    async fn get_content_args_from_latest_entry(
+        &self,
+        dataset_ref: &odf::DatasetRef,
+    ) -> Result<Option<ContentArgs>, InternalError> {
+        // TODO: Consider retractions / corrections
+        let query_res = self
+            .query_svc
+            .tail(dataset_ref, 0, 1, GetDataOptions::default())
+            .await
+            .int_err()?;
+
+        let Some(df) = query_res.df else {
+            return Ok(None);
+        };
+
+        let records = df.collect_json_aos().await.int_err()?;
+
+        assert_eq!(records.len(), 1);
+        let Value::Object(mut record) = records.into_iter().next().unwrap() else {
+            unreachable!()
+        };
+
+        // TODO: Restrict after migration
+        let content_length = usize::try_from(
+            record
+                .remove("content_length")
+                .unwrap_or_default()
+                .as_u64()
+                .unwrap_or_default(),
+        )
+        .unwrap();
+        let content_type =
+            MediaType::from(record.remove("content_type").unwrap().as_str().unwrap());
+        let content_hash = odf::Multihash::from_multibase(
+            record.remove("content_hash").unwrap().as_str().unwrap(),
```

Comment:
Same here, many references to implicit schema. What if schema columns are changed .. should the poor maintainer be looking for spread of strings all over the project?